File: killrvideo-service-statistics/src/main/java/com/killrvideo/service/statistic/repository/StatisticsRepository.java
package com.killrvideo.service.statistic.repository;

import com.killrvideo.service.statistic.dao.VideoPlaybackStatsDao;
import com.killrvideo.service.statistic.dao.VideoPlaybackStatsMapper;
import com.killrvideo.service.statistic.dto.VideoPlaybackStats;
import com.killrvideo.dse.utils.MappedAsyncPagingIterableUtils;
import org.springframework.stereotype.Repository;
import org.springframework.util.Assert;

import java.util.List;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

/**
 * Implementations of operation for Videos.
 *
 * @author DataStax Developer Advocates team.
 */
@Repository
public class StatisticsRepository {
    private final VideoPlaybackStatsDao videoPlaybackStatsDao;

    public StatisticsRepository(VideoPlaybackStatsMapper mapper) {
        this.videoPlaybackStatsDao = mapper.getVideoPlaybackStatsDao();
    }

    /**
     * Increment counter in DB (Async).
     *
     * @param videoId
     *      current videoid.
     */
    public CompletableFuture<Void> recordPlaybackStartedAsync(UUID videoId) {
        Assert.notNull(videoId, "videoid is required to update statistics");
        return videoPlaybackStatsDao.increment(videoId, 1L);
    }

    /**
     * Search for each videoid.
     *
     * @param listOfVideoIds
     *      list of EXISTING videoid
     * @return
     *      future for the list
     */
    public CompletableFuture<List<VideoPlaybackStats>> getNumberOfPlaysAsync(List<UUID> listOfVideoIds) {
        Assert.notNull(listOfVideoIds, "videoid list cannot be null");
        return videoPlaybackStatsDao.getNumberOfPlays(listOfVideoIds)
                .thenApply(MappedAsyncPagingIterableUtils::all);
    }
}

File: killrvideo-service-statistics/src/main/kotlin/com/killrvideo/service/statistic/repository/StatisticsRepository.kt
package com.killrvideo.service.statistic.repository

import com.killrvideo.dse.utils.MappedAsyncPagingIterableExtensions.all
import com.killrvideo.service.statistic.dao.VideoPlaybackStatsDao
import com.killrvideo.service.statistic.dao.VideoPlaybackStatsMapper
import com.killrvideo.service.statistic.dto.VideoPlaybackStats
import kotlinx.coroutines.future.await
import org.springframework.stereotype.Repository
import java.util.*

/**
 * Implementations of operation for Videos.
 *
 * @author DataStax Developer Advocates team.
 */
@Repository
class StatisticsRepository(mapper: VideoPlaybackStatsMapper) {
    private val videoPlaybackStatsDao: VideoPlaybackStatsDao = mapper.videoPlaybackStatsDao

    /**
     * Increment counter in DB (Async).
     *
     * @param videoId
     * current videoid.
     */
    suspend fun recordPlaybackStartedAsync(videoId: UUID): Int =
        videoPlaybackStatsDao.increment(videoId, 1L)
            .thenApply { 1 }
            .await()

    /**
     * Search for each videoid.
     *
     * @param listOfVideoIds
     * list of EXISTING videoid
     * @return
     * future for the list
     */
    suspend fun getNumberOfPlaysAsync(listOfVideoIds: List<UUID>): List<VideoPlaybackStats> =
        videoPlaybackStatsDao.getNumberOfPlays(listOfVideoIds)
            .thenApply { it.all() }
            .await()
}
